src/rust/engine/fs/src/snapshot.rs

```diff
@@ -111,22 +122,97 @@ impl Snapshot {
         directory.set_files(protobuf::RepeatedField::from_vec(files));
         store.record_directory(&directory).map(move |digest| {
           Snapshot {
-            fingerprint: digest.0,
-            digest: Some(digest),
+            digest: digest,
             path_stats: path_stats,
           }
         })
       })
       .to_boxed()
   }
+
+  pub fn contents(self, store: Arc<Store>) -> BoxFuture<Vec<FileContent>, String> {
+    Snapshot::contents_for_directory_helper(self.digest.0, store, PathBuf::from(""))
+      .map(|mut v| {
+        v.sort_by(|a, b| a.path.cmp(&b.path));
+        v
+      })
+      .to_boxed()
+  }
+
+  // Assumes that all fingerprints it encounters are valid.
+  // Returns an unsorted Vec.
+  fn contents_for_directory_helper(
+    fingerprint: Fingerprint,
+    store: Arc<Store>,
+    path_so_far: PathBuf,
+  ) -> BoxFuture<Vec<FileContent>, String> {
+    store
+      .load_directory_proto(fingerprint)
+      .and_then(move |maybe_dir| {
+        maybe_dir.ok_or_else(|| {
+          format!("Could not find directory with fingerprint {}", fingerprint)
+        })
+      })
+      .and_then(move |dir| {
+        let file_futures = join_all(
+          dir
+            .get_files()
+            .iter()
+            .map(|file_node| {
+              let path = path_so_far.join(file_node.get_name());
+              let maybe_bytes =
+                store.load_file_bytes(
+                  Fingerprint::from_hex_string(file_node.get_digest().get_hash()).unwrap(),
+                );
+              futures::future::ok(path).join(maybe_bytes)
+            })
+            .collect::<Vec<_>>(),
+        );
+        let dir_futures = join_all(
+          dir
+            .get_directories()
+            .iter()
+            .map(|dir_node| {
+              Snapshot::contents_for_directory_helper(
+                Fingerprint::from_hex_string(dir_node.get_digest().get_hash()).unwrap(),
+                store.clone(),
+                path_so_far.join(dir_node.get_name()),
+              )
+            })
+            .collect::<Vec<_>>(),
+        );
+        file_futures.join(dir_futures)
+      })
+      .and_then(
+        move |(paths_and_maybe_byteses, dirs): (Vec<(PathBuf, Option<Vec<u8>>)>,
+                                                Vec<Vec<FileContent>>)| {
+          join_all(
+            paths_and_maybe_byteses
+              .into_iter()
+              .map(|(path, maybe_bytes)| {
+                maybe_bytes
+                  .ok_or_else(|| format!("Couldn't find file contents for {:?}", path))
+                  .map(|content| FileContent { path, content })
+              })
+              .collect::<Vec<Result<FileContent, _>>>(),
+          ).join(futures::future::ok(dirs))
+        },
+      )
+      .map(|(mut files, dirs)| {
+        for mut dir in dirs.into_iter() {
```

Comment:
This all ends up doing a kind of sad amount of copying things into intermediate Vecs... Any ideas for how to avoid this are welcome!